brcm63xx: add support for Telecom Italia ADSL2+ Wi-Fi N AGPWI

DanCasterIt opened the pull request:
brcm63xx: add support for Telecom Italia ADSL2+ Wi-Fi N AGPWI

Telecom Italia ADSL2+ Wi-Fi N AGPWI a.k.a. ADB P.DG A4001N A-000-1A1-AX
is the same board as the OpenWrt's ADB P.DG A4001N with LEDs connected
to different GPIO PINs in active low configuration.

OpenWrt's ADB P.DG A4001N actually is an ADB P.DG A4001N A-000-1A1-AE

Hardware:
* SoC: Broadcom BCM6328
* RAM: 32MB
* Flash: 8MB
* Ethernet: 4x Ethernet 10/100 baseT
* Wifi 2.4GHz: Broadcom Corporation BCM43224/5 Wireless Network Adapter (rev 01)
* LEDs: 2x Power, 2x ADSL, 2x Internet, 2x Wi-Fi, 2x Service
* Buttons: 1x Reset, 1x WPS (named WiFi/LED)
* UART: 1x TTL 115200n8, TX NC RX, on J5 connector (short R192 and R193)

Installation via CFE:
* Stock CFE has to be overwriten with a Comtrend one that can upload .bin images
  with no signature check
* connect a serial port to the board
* Stop the boot process after power on by pressing enter
* set static IP 192.168.1.2
* navigate to http://192.168.1.1
* upload the OpenWrt image file

DanCasterIt committed:
brcm63xx: add support for ADB P.DG A4001N A-000-1A1-AX

ADB P.DG A4001N A-000-1A1-AX a.k.a. Telecom Italia ADSL2+ Wi-Fi N (AGPWI)
has the same PCB as the OpenWrt's ADB P.DG A4001N1 with LEDs connected
to different GPIO PINs in active low configuration.

OpenWrt's ADB P.DG A4001N image is made for the ADB P.DG A4001N A-000-1A1-AE.
It has different LEDs configuration and flash size/layout
w.r.t the ADB P.DG A4001N A-000-1A1-AX.

Hardware:
* Board ID: 96328avng
* SoC: Broadcom BCM6328
* RAM DDR2-800: 32 Mbyte - winbond W9725G6KB-25
* Serial flash: 16 Mbyte - MXIC MX25L 12845EMI-10G
* Ethernet: 4x Ethernet 10/100 baseT
* Wifi 2.4GHz: Broadcom Corporation BCM43224/5 Wireless Network Adapter (rev 01)
* LEDs: 2x Power, 2x ADSL, 2x Internet, 2x Wi-Fi, 2x Service
* Buttons: 1x Reset, 1x WPS (named WiFi/LED)
* UART: 1x TTL 115200n8, TX  NC  RX, on J5 connector (short R192 and R193)
                         NC  GND NC

Installation via CFE:
* Stock CFE has to be overwriten with one for 96328avng boards that can upload
  .bin images with no signature check (cfe-A4001N-V0000_96328avng.bin)
* connect a serial port to the board
* Stop the boot process after power on by pressing enter
* set static IP 192.168.1.2 and subnet mask 255.255.255.0
* navigate to http://192.168.1.1/
* upload the OpenWrt image file

Signed-off-by: Daniele Castro <danielecastro@hotmail.it>

adrianschmutzler commented:
commit title is too long.

adrianschmutzler commented on target/linux/brcm63xx/image/bcm63xx.mk:
```diff
@@ -207,6 +207,19 @@ endef
+define Device/AGPWI
+  $(Device/bcm63xx)
+  DEVICE_VENDOR := ADB
+  DEVICE_MODEL := P.DG A4001N A-000-1A1-AX (AGPWI)
```
Please use the DEVICE_ALT0_VENDOR/DEVICE_ALT0_MODEL syntax for the name in brackets.

DanCasterIt replied:
I corrected it. Is it ok now?

adrianschmutzler commented on target/linux/brcm63xx/image/bcm63xx.mk:
```diff
@@ -207,6 +207,19 @@ endef
+  CFE_CHIP_ID := 6328
+  FLASH_MB := 8
+  DEVICE_PACKAGES := \
+    $(USB2_PACKAGES) $(B43_PACKAGES)
```
Doesn't that fit into one line?

DanCasterIt replied:
If you refer to line 218 and 219 yes, but in the bcm63xx.mk file "$(USB2_PACKAGES) $(B43_PACKAGES)" is always in a new line after the "DEVICE_PACKAGES := \" for all the other boards.

I copied and paste these two lines from the ones in the A4001N without any modification to mantain the same code style.

Anyway if you want I put line 219 in the same line of 218

adrianschmutzler replied:
I see your point. However, I do not see any advantage in a "two-line consistency" here, and would use one line anyway if it was my PR.
Since I did only a drive-by review though, and won't merge this, feel free to choose whatever you like more.

DanCasterIt commented:
Hi @adrianschmutzler, It's been a week that I fixed the things you indicated to me.
Is there something else should I fix/do?
I'm new with pull requests :)

adrianschmutzler commented:
@DanCasterIt Device support Pull Request typically take very long until they are accepted, as reviewing them is quite a lot of work.
My remarks for this PR were only cosmetical; I'm not familiar with the target and won't review this beyond some drive-by comments. So, effectively you will have to wait until someone becomes interested in reviewing your contribution.

adrianschmutzler commented:
I've just pushed a tidy-up/reorganization patch for brcm63xx target.
For you, this would mean that you have to update device name, DTS name and the board name in board-files. I will comment inline to make this more clear...

adrianschmutzler commented on target/linux/brcm63xx/dts/agpwi.dts:
```diff
@@ -0,0 +1,125 @@
+
+/ {
+	model = "ADB P.DG A4001N A-000-1A1-AX";
+	compatible = "adb,agpwi", "brcm,bcm6328";
```
You should decide on which primary name you prefer for that device. Either use the Telecom Italia name, or the ADB name. But please don't use a different name here compared to the image/Makefile definitions. What's it with the "agpwi" BTW? I'd actually prefer if we could build a compatible based on the telecom italia model name.

If that's decide, note that you will have to consistently use that for compatible, base-files and DTS name then.

The DTS file should be named soc-vendor-model.dts, so with your current compatible it would be bcm6328-adb-agpwi.dts

DanCasterIt replied:
This router, as you understood, has three names:

the ADB one univocally identifies it and the PCB inside.

the AGPWI one univocally identifies it and the PCB inside too and it's derived from it's stock firmware name. Around the web people commonly refer to this router as AGPWI. The same exact thing was done for the Pirelli's AGPF router. OpenWrt's official images are also named as AGPF but the real Pirelli and Telecom italia's name is not AGPF, it is the Telecom Italia's firmware name for that router.

the Telecom Italia's name (now TIM) for the AGPWI does not univocally identify this router because another router commonly named as AGTWI has got the same Telecom Italia's name. These two routers are estetically different and made from two different vendors but, wird enought, the two board seems to be compatible: In the AGTWI (Technicolor TG582n) that mounts the 8MB flash (DANT-T PCB revision) you can flash the official OpenWrt image for the ADB P.DG A4001N A-000-1A1-AE (openwrt-x.x.x-brcm63xx-generic-A4001N-squashfs-cfe.bin) like for my AGPWI.
I have got the AGTWI board too but I did not flash openwrt yet and so I don't know if the image I modified for my AGPWI would work too for my AGTWI DANT-T revision or if the OpenWrt's openwrt-x.x.x-brcm63xx-generic-A4001N-squashfs-cfe.bin will work fine.
The OpenWrt's openwrt-x.x.x-brcm63xx-generic-A4001N-squashfs-cfe.bin will work for sure in both of these routers but in the AGPWI the LEDs will be wrongly mapped in the device tree and I don't know yet if my fix for the AGPWI will work for the AGTWI DANT-T too or if the actual OpenWrt image works just fine for its LEDs.

Now you can see that I'm really confused too in how to call this router in the files:
I would like to let the OpenWrt image name be AGPWI (and its internal files too for shortness)  so that people can search on the web for the AGPWI image, but I want LUCI to display the ADB name that for me is the real correct one.

So, what should I do?

adrianschmutzler replied:
Let me think about it. Just note that after my recent update the vendor name will be included in the image name, so you would have e.g. openwrt-x.x.x-brcm63xx-generic-adb_agpwi-squashfs-cfe.bin

DanCasterIt replied:
In that case the vendor name should be Telecom Italia because AGPWI is The Telecom Italia's firmware name that has notthing to do with the ADB's firmware.

I think that "openwrt-x.x.x-brcm63xx-generic-Telecom_Italia_agpwi-squashfs-cfe.bin" should be a proper image name: at least people can distinguish between Telecom Italia's AGTWI and AGPWI. Then, in the future, when I'll test this fix to the AGTWI too, I'll maybe discover that this fix works for the AGTWI as well and the image name can be changed in something like "openwrt-x.x.x-brcm63xx-generic-Telecom_Italia_ADSL2+_Wi-Fi_N-squashfs-cfe.bin" but at that point then people can't search for the AGPWI or AGPWI image anymore...
Furthermore "ADSL2+ Wi-Fi N" looks to me as a too generic name for a router...

Anyway I think I will wait for your decision about namings of this router before I apply the modifications you requested to me

DanCasterIt replied:
I think I found a solution:

I noticed that the OpenWrt image name for the AGPF is called like so because It seems that Pirelli did not give any name to that router leaving the Alice (Telecom Italia) name that is the same for both the AGPF and the AGIA router. People, to distinguish between these routers, started to use these two names AGPF and AGIA.

This is not the case for the AGPWI for wich on the bottom of the router you can find both the ADB and Telecom Italia's names.

I think I will simpli call all the files with the full ADB name like so for the dts:
"bcm6328-adb-a4001n-a-000-1a1-ax.dts"
so the OpenWrt image name will be:
"openwrt-x.x.x-brcm63xx-generic-adb-a4001n-a-000-1a1-ax-squashfs-cfe.bin"
I will not put Telecom Italia ADSL2+ Wi-Fi N AGPWI name nowhere around the sources so that people who own this router with no Telecom Italia brand name will be facilitated.

Please, let me know what do you thing of this solution :)

LuKePicci replied:
> I think I will simply call all the files with the full ADB name like so for the dts:
> "bcm6328-adb-a4001n-a-000-1a1-ax.dts"
> so the OpenWrt image name will be:
> "openwrt-x.x.x-brcm63xx-generic-adb-a4001n-a-000-1a1-ax-squashfs-cfe.bin"

That is the correct solution for sure. It would also be consistent with other newer devices from TIM which also have an AGxxx firmware name, an original marketing name, a vendor model name, and a vendor board name. For instance, "TIM Smart Modem Plus", firmware AGTEF, vendor model "DGA4130", board VBNT-K would become openwrt-x.x.x-brcm63xx-generic-technicolor-dga4130-vbnt-k-squashfs-cfe.bin, and that would also work on the same device from other service providers where no "Smart Modem Plus" or "AGTEF" names appear anywhere.

adrianschmutzler commented on target/linux/brcm63xx/dts/agpwi.dts:
```diff
@@ -0,0 +1,125 @@
+/dts-v1/;
```
Please add a licence (with SPDX identifier).

adrianschmutzler commented on target/linux/brcm63xx/base-files/lib/brcm63xx.sh:
```diff
@@ -15,6 +15,9 @@ brcm63xx_dt_detect() {
 		;;
+	"ADB P.DG A4001N A-000-1A1-AX")
+		board_name="agpwi"
+		;;
```
This is obsolete (file has been deleted).

adrianschmutzler commented on target/linux/brcm63xx/base-files/etc/board.d/01_leds:
```diff
@@ -18,6 +18,12 @@ a4001n)
 ad1018-nor)
 	ucidef_set_led_netdev "wlan0" "WLAN" "AD1018:green:wifi" "wlan0"
 	;;
+agpwi)
```
Use compatible instead.

adrianschmutzler commented on target/linux/brcm63xx/base-files/etc/board.d/02_network:
```diff
@@ -77,6 +77,7 @@ neufbox4)
 
 a4001n1 |\
 a4001n |\
+agpwi |\
```
Use compatible instead.

adrianschmutzler commented on target/linux/brcm63xx/base-files/etc/diag.sh:
```diff
@@ -15,6 +15,9 @@ set_state() {
 	ad1018-nor)
 		status_led="AD1018:green:power"
 		;;
+	agpwi)
```
Use compatible instead.

adrianschmutzler commented on target/linux/brcm63xx/base-files/etc/uci-defaults/09_fix_crc:
```diff
@@ -13,6 +13,7 @@ do_fixcrc() {
 case "$(board_name)" in
 	a4001n |\
 	a4001n1 |\
+	agpwi |\
```
Use compatible instead.

adrianschmutzler commented on target/linux/brcm63xx/dts/agpwi.dts:
```diff
@@ -0,0 +1,125 @@
+};
+
+&uart0 {
+	status = "ok";
```
I think "okay" is prefered here.

DanCasterIt replied:
I copied the content of this file from the A4001N.dts and just modified GPIO stuff, that's where this "ok" comes from. I left "ok" for code style uniformity with the tween PCB A4001N.

Do you steel want me to change "ok" into "okay"?

adrianschmutzler replied:
okay is preferred, while ok also works. Since we add something new here, please update all ok to okay in the DTS.

adrianschmutzler commented on target/linux/brcm63xx/image/bcm63xx.mk:
```diff
@@ -234,6 +234,21 @@ define Device/A4001N1
+  DEVICE_MODEL := ADSL2+ Wi-Fi N AGPWI
+  DEVICE_ALT0_VENDOR := ADB
+  DEVICE_ALT0_MODEL := P.DG A4001N A-000-1A1-AX
+  DEVICE_DTS := agpwi
```
This line is obsolete if you name device and DTS file correctly.

adrianschmutzler commented on target/linux/brcm63xx/image/bcm63xx.mk:
```diff
@@ -234,6 +234,21 @@ define Device/A4001N1
+  CFE_CHIP_ID := 6328
+  FLASH_MB := 8
+  DEVICE_PACKAGES := \
+    $(USB2_PACKAGES) $(B43_PACKAGES)
```
This will match into one line.

DanCasterIt replied:
I copied and paste  this too from the A4001N rows and left them as they are for code style uniformity.
Do you still want me to match it in one line?

adrianschmutzler replied:
I've just updated this and converted all to a single line.

adrianschmutzler commented on target/linux/brcm63xx/patches-4.14/805-board_agpwi.patch:
```diff
@@ -0,0 +1,68 @@
+--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
++++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
```
Please also provide a patch for kernel 4.19.

DanCasterIt commented:
@adrianschmutzler I added all the modifications you requested and I named averything with the ADB name of this router since there's people who own this router with no brand and/or with other brands.

The ADB name should be always present on the bottom of the device like in my case.

For this reason I changed the name of the router in the commit/PR title too.

adrianschmutzler commented on target/linux/brcm63xx/image/bcm63xx.mk:
```diff
@@ -215,6 +215,17 @@ define Device/adb_a4001n1
 endef
 TARGET_DEVICES += adb_a4001n1
 
+define Device/adb_a4001n-a-000-1a1-ax
```
I like that much more, but if we go the way of terribly long (but accurate) names anyway, the pdg should be included as well (pdg or p-dg).

Despite, if I understand you correctly, the current a4001n existing in OpenWrt is not precisely named at the moment, as it's actually a subspecies of the a4001n. Since I just recently changed the target names in brcm63xx anyway, I'm tempted to _correct_ that right away. Thoughts? And what should happen to a4001n1 then?

DanCasterIt replied:
I understand your point: it was the reason why I originally called it AGPWI in the files instead of its long ADB name.

The guy who added support for the ADB P.DG A4001N A-000-1A1-AE called the image "A4001N" for this reason too: for shortness. He most probably thought that "A4001N" could be more than enought to univocally identify this PCB, but it was not true.

To answer your question yes, "A4001N" should be corrected in "ADB P.DG A4001N A-000-1A1-AE". I was planning to do a PR for that. The same thing as to be done for the "A4001N1", this is not its full name but I did not search yet its full name.

I'm pretty sure that "P.DG" is just the full name of ADB company, not something userfull to univocally identify the PCB like "A-000-1A1-AE". I would not add this name too for shortness.

DanCasterIt replied:
I was wrong: P.DG identifies the board to, but I'm pretty sure it's not needed to univocally identify the PCB, so I would not add it for shortnessa

adrianschmutzler replied:
I look at this from a different angle: Many problems with device names are created because we try to think how we could make names shorter and what we think will create unique names at the present time. Then, some new unanticipated device comes up and our great plan of uniqueness is destroyed.
Thanks why I initially motivated you to take the full name for the device string as well, as this is the easiest and cleanest approach, with the only disadvantage of having long device names.
Despite, consistency between model/DEVICE_MODEL and the image name will also help users looking for a device. Thus, I would still prefer to add the pdg after all. When I google for "A4001N", the P.DG is always there, so it's part of the name after all.

DanCasterIt replied:
Ok, so you want me to add "pdg" (or "p-dg" ?) to all the file names in the sources?

adrianschmutzler replied:
I think the only file name would be the DTS, but I just think of
adb_p-dg-a4001n-a-000-1a1-ax
or
adb_pdg-a4001n-a-000-1a1-ax
(or with comma for the compatible). Won't make it much worse btw.

The only thing I'm not sure about yet is the led labels, but if I stick to my own concept they also need the long name (without vendor).

DanCasterIt replied:
ok, so you want me to change the DTS name only? I think I will use "adb_pdg-a4001n-a-000-1a1-ax".

The real problem with LEDs, in my opinion, is that their long name "A4001N_A-000-1A1-AX:green:internet" is shown il LUCI too. To me it has not to be shown. It should be just "green:internet"

adrianschmutzler replied:
I want to have compatible, model name, DEVICE_MODEL, device node name, DTS file name, and LED model prefix with the same "name", e.g. with pdg in this case.
The LED thing is very annoying indeed, unfortunately having the model there seems to be a (though only formal) kernel requirement, I also do not see a practical use in it.

I'd prefer lower-case and hyphens only for the LED prefix, that's typical at other more modern targets, as it allows to generate the prefix from device node name:
adb_pdg-a4001n-a-000-1a1-ax -> "cut at underscore" -> pdg-a4001n-a-000-1a1-ax

Then, you do not have to put the long board names in base-files explicitly, but can just use a variable, e.g. https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/generic/base-files/etc/board.d/01_leds#L12

adrianschmutzler commented:
As it looks (and as you stated), the only difference to the other a4001n are the LEDs. Thus, it would make sense to introduce a shared DTSI for both of them:
bcm6328-adb-pdg-a4001n.dtsi
or
bcm6328-adb-p-dg-a4001n.dtsi

And you also do not need to define a new board_info with the same content either, just add
{ .compatible = "adb,a4001n-a-000-1a1-ax", .data = &board_A4001N, },
to the 566-board_A4001N.patch

So, after all, I'd do the following in this PR:
commit 1: Update the names for "old" A4001N (compatible, image, dts file name, compatible in board patch, but not the name of the board_info)
commit 2: Add "your" A4001N on top, create dtsi, extend board patch with the new compatible.

Sorry for making it this complicated, but I think it will be a better implementation afterwards then.

DanCasterIt commented:
Is there something similar that was done for other boards that I can use as example to make these modifications?

maybe in the brcm63xx DTS folder?

adrianschmutzler commented:
Just wait one day or two, I will quickly do the adjustments ("commit 1") myself and then you can pull them into your PR and rebase your stuff on it.

DanCasterIt commented:
@adrianschmutzler ok, thank you.

Just one more doubt: will this new aproach work too if I'll discover that my patch for ADB P.DG A4001N A-000-1A1-AX works also for Technicolor TG582n  DANT-T rev. (AGTWI)?
It has both different vendor and model but runs the "A4001N" image too.

Furthermore there's also the Technicolor TG582n  DANT-1 rev. (still AGTWI) that runs the "A4001N1" image instead

adrianschmutzler commented:
> @adrianschmutzler ok, thank you.
> 
> Just one more doubt: will this new aproach work if I'll discover that my patch for ADB P.DG A4001N A-000-1A1-AX works also for Technicolor TG582n DANT-T rev. (AGTWI)? I think that could be possible basing of the fact that they have the same number of LEDs with the same name and are distributer with the same Telecom Italia name.
> 
> It has both different vendor and model w.r.t. the AGPWI but runs the "A4001N" image too like the AGPWI.
> 
> Furthermore there's also the Technicolor TG582n DANT-1 rev. (still AGTWI) that runs the "A4001N1" image instead
> 
> These two images, a part from LEDs, seems to differ just on the flash size:
> https://openwrt.org/toh/thomson/tg582n

Well, how would you solve it? To me, it looks like the best way to get an overview over this mess is to add the different device names, even if this produces a few clones.

adrianschmutzler commented:
The more I discuss this, the more I think I should have kept my hands out of brcm63xx in the first place...

DanCasterIt commented:
> 
> 
> > @adrianschmutzler ok, thank you.
> > Just one more doubt: will this new aproach work if I'll discover that my patch for ADB P.DG A4001N A-000-1A1-AX works also for Technicolor TG582n DANT-T rev. (AGTWI)? I think that could be possible basing of the fact that they have the same number of LEDs with the same name and are distributer with the same Telecom Italia name.
> > It has both different vendor and model w.r.t. the AGPWI but runs the "A4001N" image too like the AGPWI.
> > Furthermore there's also the Technicolor TG582n DANT-1 rev. (still AGTWI) that runs the "A4001N1" image instead
> > These two images, a part from LEDs, seems to differ just on the flash size:
> > https://openwrt.org/toh/thomson/tg582n
> 
> Well, how would you solve it? To me, it looks like the best way to get an overview over this mess is to add the different device names, even if this produces a few clones.

Yes, I would do the same thing you described. I told you about the AGTWI in advance just to be sure that for the software architecture design you are going to make this will not represent a big issue.

Anyway If I well understood flash size and partition ofsets are also stored in the DTS. If so, "A4001N" and "A4001N1" differ just for the memory size (this is what the link I shared with you reports). So there solud be kind of a generic "A4001*" that can be built into 3 different devices (5 with the two AGTWIs)

Furthermore in the kernel (in the board_info) and in the make script I left .name = "96328dg2x2" as it was in the "A4001N" but I really don't know where to search the correct one for my router. "A4001N1" has .name = "963281T_TEF". Can this represent a problem in unifiing everything in a single dts file? Is this name so important? what is it?

DanCasterIt commented:
> 
> 
> The more I discuss this, the more I think I should have kept my hands out of brcm63xx in the first place...

At least now I have someone who can review my PR :)
I think ath79 has other kind or similar issues too ;)

DanCasterIt commented:
@adrianschmutzler I discovered that there are many more clones of the "a4001n" board than I imagin for example the "ADB P.DG A4101N" seems to run the "a4001n" image too...

At this point I think that even if these boards have original board ID diferent from the one of the ADB P.DG A4001N A-000-1A1-AE it could be a good idea, in terms of order, to unify all the these boards under an single DTS. Even tho these boards have different board ID names. I discovered that OpenWrt seems to do not check for the board ID during boot time for the brcm63xx architecture. So even if the ID is wrong, it the image boots anyway.

Sorry for the confusion but I think you can understand my reasons

adrianschmutzler commented:
I'm not opposed to having a shared image for several devices if it has a reasonable naming scheme. So far, however, I do not see one. I'm still struggling what's the best solution here.
Anyway, please don't waste your time on rebasing or renaming until we have a conclusion here, I will do that when we agreed on what's the way forward here.

adrianschmutzler commented:
> If so, "A4001N" and "A4001N1" differ just for the memory size

They differ in flash size: 8 MiB vs. 16 MiB.

https://github.com/openwrt/openwrt/blob/master/target/linux/bcm63xx/dts/bcm6328-adb-a4001n.dts#L91
https://github.com/openwrt/openwrt/blob/master/target/linux/bcm63xx/dts/bcm6328-adb-a4001n1.dts#L122

If it wasn't for the LEDs, we could just name them a4001n-8m and a4001n-16m...

DanCasterIt commented:
@adrianschmutzler now I think I have a clair idea:

We do not have any reasonable naming scheme because the same board has been shared between many vendors and each one of them gave a different name and board ID to the router changing memory size and/or LED configuration.

Meanwhile we are discussing here I'm discussing of this PR in an Italian forum. The same forum where the AGPWI and many other ADB and Technicolor routers were unlocked.

I discovered that people use to flash both "a4001n" and "a4001n1" just to see which one as more LEDs blinking properly. This let me think that, despite the different memory size, boot address and nvram offset remain the same. Otherwise I can't explain why a 16m image works in a 8m router too...

till now I discovered that:
"A4001N" image runs in:
ADB P.DG A4001N                           (from movistar in Spain)
ADB P.DG A4001N A-000-1A1-AX   (AGPWI from Telecom Italia)
ADB P.DG A4001N A-000-1A1-AE   (from Arnet, Telecom Argentina, OpenWrt's actual image)
ADB P.DG A4101N A-000-1A1-AE   (from TELETU in Italy)
Technicolor TG582n DANT-T            (AGTWI from Telecom Italia)

"A4001N1" image runs in:
ADB P.DG A4001N1                         (from movistar in Spain, OpenWrt's actual image)
Technicolor TG582n DANT-1            (AGTWI from Telecom Italia)

but in the forum they told me that these two OpenWrt images run in really a lot of devices. So, as you can se, "A4001N" and "A4001N" compatible devices are poppoing out like mushrooms meanwhile I discuss, each one of them with different board ID. So, tecnically, each board is different from the other and can't be considered ad a subspecies.

So after all we basically have two options:
A: we keep the actual DTS file structure
B: we make a DTSI trying to unify as meny PCBs as possibile

Option A, for now, lets us merge 3 out of 5 devices counting the future ones too:
ADB P.DG A4001N                           (from movistar in Spain)
ADB P.DG A4001N A-000-1A1-AX   (AGPWI from Telecom Italia)
ADB P.DG A4001N A-000-1A1-AE   (from Arnet, Telecom Argentina, OpenWrt's actual image)
but in my opinion option A is not formally correct because these board (most probably) have different board IDs and we don't really know if they really differ just for LEDs and flash size because if you look at some pictures they differ in the PCB too and we don't know if it's just a phisical difference or if there are also other differences a part from flash size and LED names. All we know is that OpenWrt's "A4001N" runs in these devices. So, in few words, if we go for this option we decided to don't kare about different board ID and eventually future discoverings about further differences

Option B is formally correct with the only disadvantage that we can't merge 3 clones

Further more I discovered today that my AGPWI as a 128 Mb flash instead of an 8Mb one...
I actually don't know my AGPWI board ID from the stock CFE, but I can put one printed on the PCB (at least for now) that is "IO1L061T00 LF". When I'll put my hands on a brand new AGPWI I will upload the right board ID if this is not good.

adrianschmutzler commented:
> I discovered that people use to flash both "a4001n" and "a4001n1" just to see which one as more LEDs blinking properly.

And that's exactly why we need device-specific images. It's just a coincidence that flashing the wrong flash size does not brick the device on this platform. Despite, I consider "device support" to include the specific LED configuration. If LEDs/GPIOs are different, it's a different device and needs different support.

> ADB P.DG A4001N (from movistar in Spain)
> ADB P.DG A4001N A-000-1A1-AX (AGPWI from Telecom Italia)
> ADB P.DG A4001N A-000-1A1-AE (from Arnet, Telecom Argentina, OpenWrt's actual image)

Do you know whether the movistar device also has some specific number?

For me, this sounds like the following at the moment:

I create two dtsi: adb_pdg-a4001n-8m.dtsi and adb_pdg-a4001n-16m.dtsi.

Those will contain the shared stuff including partitioning.

The LED configuration will be moved to the device-specific DTS files.

Then, as you state that the current A4001N is actually A-000-1A1-AE, I will rename the device and the dts accordingly, and derive from a4001n-8m.dtsi.

I'd then put your patch for A-000-1A1-AX on top, and derive from a4001n-8m for now, until we now better about partitioning.

I would only add the other devices mentioned if somebody will retrieve the correct LED GPIOs and tests on them then. But each of these devices should get it's own image, otherwise we will never be clear to outsiders.

adrianschmutzler commented:
> because the same board has been shared between many vendors

To make my position a little more understandable: I do not think in _boards_, I do think in _devices_. If the same board is used in two different devices, they should get different images (as the users sees the device, not the board; and the device might include different peripherals to the same board).

DanCasterIt commented:
> 
> 
> > I discovered that people use to flash both "a4001n" and "a4001n1" just to see which one as more LEDs blinking properly.
> 
> And that's exactly why we need device-specific images. It's just a coincidence that flashing the wrong flash size does not brick the device on this platform. Despite, I consider "device support" to include the specific LED configuration. If LEDs/GPIOs are different, it's a different device and needs different support.
> 
> > ADB P.DG A4001N (from movistar in Spain)
> > ADB P.DG A4001N A-000-1A1-AX (AGPWI from Telecom Italia)
> > ADB P.DG A4001N A-000-1A1-AE (from Arnet, Telecom Argentina, OpenWrt's actual image)
> 
> Do you know whether the movistar device also has some specific number?

no, it's just ADB P.DG A4001N with notthing more.

> 
> For me, this sounds like the following at the moment:
> 
> I create two dtsi: adb_pdg-a4001n-8m.dtsi and adb_pdg-a4001n-16m.dtsi.
> 

I don't know if there exists a4001n routers with 16m of flash. Probably yes but I don't know. I know that a4001n1 has 16m.

> Those will contain the shared stuff including partitioning.
> 
> The LED configuration will be moved to the device-specific DTS files.
> 
> Then, as you state that the current A4001N is actually A-000-1A1-AE, I will rename the device and the dts accordingly, and derive from a4001n-8m.dtsi.
> 
> I'd then put your patch for A-000-1A1-AX on top, and derive from a4001n-8m for now, until we now better about partitioning.
> 

even tho I discovered today that flash size of A-000-1A1-AX is 128Mb instead of 8Mb?
I updated my comment adding this too. Sorry for that. Did you see it?

> I would only add the other devices mentioned if somebody will retrieve the correct LED GPIOs and tests on them then. But each of these devices should get it's own image, otherwise we will never be clear to outsiders.

DanCasterIt commented:
> 
> 
> > because the same board has been shared between many vendors
> 
> To make my position a little more understandable: I do not think in _boards_, I do think in _devices_. If the same board is used in two different devices, they should get different images (as the users sees the device, not the board; and the device might include different peripherals to the same board).

Ok, for me board = PCM. So, A-000-1A1-AX and A-000-1A1-AE have:
- different phisical PCB layout
- totally different chasis (is this a different device for you?)

if so, if I well understood, they should have two different images. The same thing applyes for all the router models I listed in the previous comment. They all externally look totally different

DanCasterIt commented:
@adrianschmutzler with bcm6328-adb-a4001n-a-000-1a1-ax.dts including bcm6328-adb_pdg-a4001n-128m.dtsi can I put a different board ID?

In the kernel there's a wrong board ID:
A-000-1A1-AX has 
"IO1L061T00 LF" with no external wifi antennas and is newer (the one I own)
"IO1L063T10 LF" with external wifi antennas and is older (I have to check if it as a different flash size)

A-000-1A1-AX has 
"96328dg2x2" both in the kernel sources and in the DTS files

DanCasterIt commented:
> 
> 
> @adrianschmutzler now I think I have a clair idea:
> 
> We do not have any reasonable naming scheme because the same board has been shared between many vendors and each one of them gave a different name and board ID to the router changing memory size and/or LED configuration.
> 
> Meanwhile we are discussing here, I'm discussing of this PR in an Italian forum. The same forum where the AGPWI and many other ADB and Technicolor routers were unlocked.
> 
> I discovered that people use to flash both "a4001n" and "a4001n1" just to see which one as more LEDs blinking properly. This let me think that, despite the different memory size, boot address and nvram offset remain the same. Otherwise I can't explain why a 16m image works in a 8m router too...
> 
> till now I discovered that:
> "A4001N" image runs in:
> ADB P.DG A4001N (from movistar in Spain)
> ADB P.DG A4001N A-000-1A1-AX (AGPWI from Telecom Italia)
> ADB P.DG A4001N A-000-1A1-AE (from Arnet, Telecom Argentina, OpenWrt's actual image)
> ADB P.DG A4101N A-000-1A1-AE (from TELETU in Italy)
> Technicolor TG582n DANT-T (AGTWI from Telecom Italia)
> 
> "A4001N1" image runs in:
> ADB P.DG A4001N1 (from movistar in Spain, OpenWrt's actual image)
> Technicolor TG582n DANT-1 (AGTWI from Telecom Italia)
> 
> but in the forum they told me that these two OpenWrt images run in really a lot of devices. So, as you can see, the more I discuss about "A4001N" and "A4001N" compatible devices, the more are poppoing out new ones like mushrooms, each one of them with different board ID. So, tecnically, each board is different from the other and can't be considered as a subspecies.
> 
> So after all we basically have two options:
> A: we keep the actual DTS file structure
> B: we make a DTSI trying to unify as many PCBs as possibile
> 
> Option A, for now, allows us to merge 3 out of 5 devices counting the future ones too:
> ADB P.DG A4001N (from movistar in Spain)
> ADB P.DG A4001N A-000-1A1-AX (AGPWI from Telecom Italia)
> ADB P.DG A4001N A-000-1A1-AE (from Arnet, Telecom Argentina, OpenWrt's actual image)
> but in my opinion option A is not formally correct because these board (most probably) have different board IDs and we don't really know if they differ just for LEDs and flash size because if you look at some pictures they differ in the PCB too and we don't know if it's just a phisical difference or if there are further differences a part from flash size and LED names. All we know is that OpenWrt's "A4001N" runs in these devices. So, in few words, if we go for this option we decide to don't care about different board ID and eventually future discoverings about further differences
> 
> Option B is formally correct with the only disadvantage that we can't merge 3 clones
> 
> Further more I discovered today that my AGPWI flash has 128Mb instead of 8Mb...
> I actually don't know my AGPWI board ID from the stock CFE, but I can put one printed on the PCB (at least for now) that is "IO1L061T00 LF". When I'll put my hands on a brand new AGPWI I will upload the right board ID if this is not good.
> 
> More precisely there are two models of ADB P.DG A4001N A-000-1A1-AX (AGPWI from Telecom Italia):
> "IO1L061T00 LF" has no external wifi antennas and is newer (the one I own)
> "IO1L063T10 LF" has external wifi antennas and is older (I have to check if it as a different flash size)

@adrianschmutzler can you please have a look again to this? I think edited it many times meanwhile you were answering me. Sorry.

adrianschmutzler commented:
> in both the kernel (board_bcm963xx.c) and the make (bcm63xx.mk) there's a wrong board ID:
> A-000-1A1-AX has
> "IO1L061T00 LF" model with no external wifi antennas and is newer (the one I own)
> "IO1L063T10 LF" model with external wifi antennas and is older (I have to check if it has a different flash size)

Well then, so we got ourselves two variant of the A-000-1A1-AX that need different images ...

DanCasterIt commented:
> 
> 
> > in both the kernel (board_bcm963xx.c) and the make (bcm63xx.mk) there's a wrong board ID:
> > A-000-1A1-AX has
> > "IO1L061T00 LF" model with no external wifi antennas and is newer (the one I own)
> > "IO1L063T10 LF" model with external wifi antennas and is older (I have to check if it has a different flash size)
> 
> Well then, so we got ourselves two variant of the A-000-1A1-AX that need different images ...

a part from the board ID I don't know if they really have different flash size (around the web there's notthing about this). The only way to konow it is to open the device. Here with me I have only the one with no antennas (IO1L061T00 LF).
They have two PCBs differing just for the antennas. In the model name there's no difference. It's still ADB P.DG A4001N A-000-1A1-AX for both

adrianschmutzler commented:
You will have to open the device anyway to determine the correct partitioning ...

DanCasterIt commented:
> 
> 
> You will have to open the device anyway to determine the correct partitioning ...

Most probably not: around the web people use to refer to these two routers as it were one. That's probably the reason why they let you download only one image called "AGPWI_1.1.0_013.bin" to restore the stock firmware. This image seems to work for both (it was manually made from the same people who unlocked the router).

I told you that there can there can be a different flash size, both because there's a lak of informations on the flash size around the web (I had to open the router today), and because they did the same thing with AGPWI with DANTs revisions.

even the board ID, on the PCB is different (I'm guessing this is the board ID) between the two... but it could be that in the stock CFE it will be an unic one for both since it's choosen at compile time...

For these reasons I suggest to add support for A-000-1A1-AX as it were one and put as board ID something like "IO1L06*T*0 LF" in the bcm63xx.mk and board_bcm963xx.c files. Just to differenciate it from the other ones. When I will put my hands on a new AGPWI with the stock CFE I will update the right board ID taken from the stock CFE

DanCasterIt commented:
Another possibility is to neglect the board ID and unify everything under the 96328dg2x2 board ID of the A-000-1A1-AE.

OpenWrt does not check for the board ID at boot time. So this approach will cause no issues...

actually I'm running an image with the wrong board ID and wrong flash size but it works... its been working for years

even if the routers have different flash size we saw that they do not brik. The aproach of making an unique image for A-000-1A1-AX seems to do not cause any damage...

adrianschmutzler commented:
That's exactly how device support in OpenWrt not works. It's not like "we put something together that somehow works because it worked for so many years before".

In contrast, please verify what your device actually is. Then create a commit that will support this very board by putting together all the precise information about it you collected before. This will not mean relying on some image created by some people nobody knows, but on reliable information on the device, like a bootlog during serial connection or photos of the board showing it's components. This will mean putting the _correct_ board name and not just something. This will mean putting the correct partitioning scheme into the DTS. Just putting together information from the internet is not enough.

DanCasterIt commented:
@adrianschmutzler I flashed the stock firmware to my AGPWI (from the backup I did with dd before unlocking it of all 16Mb of flash) and now I know for sure that the device id is 96328avng.

I did the same thing with my AGPWI with antennas and I discovered that partition layout, flash size and board ID are exactly the same betheewn these two models.

DanCasterIt commented:
> 
> 
> I want to have compatible, model name, DEVICE_MODEL, device node name, DTS file name, and LED model prefix with the same "name", e.g. with pdg in this case.
> The LED thing is very annoying indeed, unfortunately having the model there seems to be a (though only formal) kernel requirement, I also do not see a practical use in it.
> 
> I'd prefer lower-case and hyphens only for the LED prefix, that's typical at other more modern targets, as it allows to generate the prefix from device node name:
> adb_pdg-a4001n-a-000-1a1-ax -> "cut at underscore" -> pdg-a4001n-a-000-1a1-ax
> 
> Then, you do not have to put the long board names in base-files explicitly, but can just use a variable, e.g. https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/generic/base-files/etc/board.d/01_leds#L12
> 
> You see, my agenda is all about consistency here. Despite, lower-case strings will be shorter.

@adrianschmutzler I modified diag.sh and 01_leds like in the example link you linked.

I named the DTS "bcm6328-adb_pdg-a4001n-a-000-1a1-ax.dts" and had to add
"DEVICE_DTS := bcm6328-adb_pdg-a4001n-a-000-1a1-ax" to bcm63xx.mk otherwise it searches for a non existing DTS file name called "bcm6328-adb-pdg-a4001n-a-000-1a1-ax.dts"

adrianschmutzler commented:
> I named the DTS "bcm6328-adb_pdg-a4001n-a-000-1a1-ax.dts" and had to add
"DEVICE_DTS := bcm6328-adb_pdg-a4001n-a-000-1a1-ax" to bcm63xx.mk otherwise it searches for a non existing DTS file name called "bcm6328-adb-pdg-a4001n-a-000-1a1-ax.dts"

You will have to use "bcm6328-adb-pdg-a4001n-a-000-1a1-ax.dts" for the DTS name then. The use of "-" and "_" is inconsistent between targets, so I may have given wrong advice here before.

DanCasterIt commented:
> You will have to use "bcm6328-adb-pdg-a4001n-a-000-1a1-ax.dts" for the DTS name then. The use of "-" and "_" is inconsistent between targets, so I may have given wrong advice here before.

ok, I made the modifications:
- Changed the DTS name in "bcm6328-adb-pdg-a4001n-a-000-1a1-ax.dts"
- removed the redundant string in the bcm63xx.mk file

Noltari commented:
Merged with some modifications d6f66dd88e5d833f88fa4b6e056499ec7644ff7c:
- LEDs exposed with controller instead of gpios.
- Updated to current OpenWrt master.

Thanks!
